### PR TITLE
Expunge Vector

### PR DESCRIPTION
**Pull request by connorjward**
This has been deprecated for a very long time.

### CODE REVIEW

File: pyop2/types/dat.py [hunk 1]
```diff
@@ -242,6 +242,18 @@ def data_wo_with_halos(self):
...
+    @mpi.collective
+    def global_data(self):
+        with self.vec_ro as vec:
+            v = PETSc.Vec().createSeq(vec.size, comm=mpi.COMM_SELF)
+            is_ = PETSc.IS().createStride(vec.size, 0, 1, comm=mpi.COMM_SELF)
+
+            vscat = PETSc.Scatter().create(vec, is_, v, None)
+            vscat.scatterBegin(vec, v, addv=PETSc.InsertMode.INSERT_VALUES)
+            vscat.scatterEnd(vec, v, addv=PETSc.InsertMode.INSERT_VALUES)
+        return v.array
```
Comment (JHopeCollins) [inline pyop2/types/dat.py]: This can be written a little more simply, see:
https://github.com/firedrakeproject/firedrake/blob/3f86c911d69e46e8b2607a4c18e0674c68189024/firedrake/adjoint_utils/ensemble_function.py#L33-L39

Comment (connorjward) [reply pyop2/types/dat.py]: Awesome thank you. I just copied what was there.

Comment (connorjward) [reply pyop2/types/dat.py]: I've done this but I think there is a bug here. `toAll()` returns a tuple so calling `.scatter` is undefined. https://petsc.org/main/petsc4py/reference/petsc4py.PETSc.Scatter.html#petsc4py.PETSc.Scatter.toAll

Comment (JHopeCollins) [reply pyop2/types/dat.py]: ~API must have changed. Looks like it now returns both the scatter and an appropriate Vec, so you could probably do something like this:~ Never mind, you already did!
```python
scatter, lvec = PETSc.Scatter().toall(gvec)
scatter.scatter(gvec, lvec, addv=PETSc.InsertMode.INSERT_VALUES) 
```

File: .github/workflows/core.yml [hunk 2]
```diff
@@ -336,7 +336,8 @@ jobs:
           matrix.arch == 'default'
         run: |
           . venv/bin/activate
-          git clone --depth 1 https://github.com/thetisproject/thetis.git thetis-repo
+          : # UNDO ME
+          git clone --depth 1 https://github.com/thetisproject/thetis.git thetis-repo --branch connorjward/novector
```
Comment (connorjward) [inline .github/workflows/core.yml]:
```suggestion
          git clone --depth 1 https://github.com/thetisproject/thetis.git thetis-repo
```

File: pyop2/types/dat.py [hunk 3]
```diff
@@ -242,6 +242,16 @@ def data_wo_with_halos(self):
...
 
+    @property
+    @mpi.collective
+    def global_data(self):
+        """Return all the data for the Dat gathered onto individual ranks."""
+        with self.vec_ro as gvec:
+            scatter, lvec = PETSc.Scatter().toAll(gvec)
+            scatter.scatter(
+                gvec, lvec, addv=PETSc.InsertMode.INSERT_VALUES)
+        return lvec.array_r
```
Comment (JHopeCollins) [inline pyop2/types/dat.py]: Does this need to be the read-only accessor?

Comment (connorjward) [reply pyop2/types/dat.py]: Fair point. Fixed.

File: demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst [hunk 4]
```diff
@@ -287,11 +287,11 @@ To save data for visualization, we change the position of the nodes in the mesh,
         output_data.counter += 1
         if output_data.counter % output_data_every_x_time_steps != 0:
             return
-        mesh_static = mesh.coordinates.vector().get_local()
-        mesh.coordinates.vector().set_local(mesh_static + X.vector().get_local())
-        mesh.coordinates.dat.data[:, 1] += eta.dat.data_ro
+        mesh_static = mesh.coordinates.copy(deepcopy=True)
+        mesh.coordinates += X
+        mesh.coordinates.dat.data_wo[:, 1] += eta.dat.data_ro
```
Comment (JHopeCollins) [inline demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst]: Is `dat.data_wo` correct if you are _incrementing_ the values rather than _overwriting_ them? Should it not be `dat.data`?

File: demos/saddle_point_pc/saddle_point_systems.py.rst [hunk 5]
```diff
@@ -93,13 +93,11 @@ discontinuous space. ::
...
 
-To initialise this function to a random value we access its :class:`~.Vector`
-form and use numpy_ to set the values::
+We initialise this function to hold random values::
 
     #
         import numpy as np
-        fvector = f.vector()
-        fvector.set_local(np.random.uniform(size=fvector.local_size()))
+        f.dat.data_wo[...] = np.random.uniform(size=f.dof_dset.size)
```
Comment (JHopeCollins) [inline demos/saddle_point_pc/saddle_point_systems.py.rst]: Would `f.dat.data_wo[...] = np.random.uniform(size=f.dat.data_wo.shape)` not be the obvious pattern?

File: firedrake/adjoint_utils/function.py [hunk 6]
```diff
@@ -315,25 +308,23 @@ def _ad_function_space(self, mesh):
...
     def _applyBinary(self, f, y):
-        vec = self.vector()
-        npdata = vec.get_local()
-        npdatay = y.vector().get_local()
+        npdata = self.dat.data_ro
+        npdatay = y.dat.data_ro
         for i in range(len(npdata)):
             npdata[i] = f(npdata[i], npdatay[i])
-        vec.set_local(npdata)
+        self.dat.data_wo[...] = npdata
```
Comment (JHopeCollins) [inline firedrake/adjoint_utils/function.py]: `npdata` points to the read-only array on line 323, but is written into on line 326. Is doing the right thing?
If you just write into `self.dat.data_wo` directly then you can also remove the additional copy on line 327.

File: tests/firedrake/adjoint/test_assemble.py [hunk 7]
```diff
@@ -99,7 +95,7 @@ def test_assemble_1_forms_tlm():
 
     Jhat = ReducedFunctional(J, Control(f))
     h = Function(V)
-    h.vector()[:] = rand(h.dof_dset.size)
+    h.dat.data_wo[...] = rand(h.dof_dset.size)
```
Comment (JHopeCollins) [inline tests/firedrake/adjoint/test_assemble.py]: We should be consistent with how we are querying the array size for `dat.data`.

1. `dat.data.shape`?
2. `h.dof_dset.size`?
3. `V.dim()`?

Comment (connorjward) [reply tests/firedrake/adjoint/test_assemble.py]: Sure. Note that the last of these isn't safe because that gives the global size.

Comment (JHopeCollins) [reply tests/firedrake/adjoint/test_assemble.py]: Thanks. The last one is used in quite a few places so even more reason to make it consistent. I think it's only used in serial tests but we don't want people copying unsafe code.

File: demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst (hunk 4 above)
Comment (connorjward) [reply demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst]: Good catch

File: firedrake/adjoint_utils/function.py (hunk 6 above)
Comment (connorjward) [reply firedrake/adjoint_utils/function.py]: done

File: tests/firedrake/adjoint/test_assemble.py (hunk 7 above)
Comment (connorjward) [reply tests/firedrake/adjoint/test_assemble.py]: done

File: demos/saddle_point_pc/saddle_point_systems.py.rst (hunk 5 above)
Comment (connorjward) [reply demos/saddle_point_pc/saddle_point_systems.py.rst]: done

File: demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst (hunk 4 above)
Comment (connorjward) [reply demos/linear_fluid_structure_interaction/linear_fluid_structure_interaction.py.rst]: done

**Review by JHopeCollins — APPROVED** [review]
LGTM. Thanks for taking this on!

File: tests/firedrake/adjoint/test_hessian.py [hunk 8]
```diff
@@ -179,7 +174,7 @@ def test_nonlinear():
     Jhat = ReducedFunctional(J, Control(f))
 
     h = Function(V)
-    h.vector()[:] = 10*rng.random(V.dim())
+    h.dat.data_wo[...] = 10*rng.random(h.dat.data_wo.shape)
```
Comment (pbrubeck) [inline tests/firedrake/adjoint/test_hessian.py]: We should make these tests deterministic
```suggestion
    rg = RandomGenerator(PCG64(seed=1234))
    h = rg.uniform(V, 0, 10)
```
Also the line above should be deleted

Comment (connorjward) [reply tests/firedrake/adjoint/test_hessian.py]: Done

File: firedrake/ml/pytorch/fem_operator.py [hunk 9]
```diff
@@ -220,7 +217,6 @@ def from_torch(x, V=None):
             val = val[0]
         return Constant(val)
     else:
-        x = x.detach().numpy()
         x_F = Function(V)
```
Comment (pbrubeck) [inline firedrake/ml/pytorch/fem_operator.py]:
```suggestion
        x_F = Function(V, val=x.detach().numpy())
```

File: firedrake/ml/jax/fem_operator.py [hunk 10]
```diff
@@ -225,7 +222,6 @@ def from_jax(x: "jax.Array", V: Optional[WithGeometry] = None) -> Union[Function
             val = val[0]
         return Constant(val)
     else:
-        x = np.asarray(x)
         x_F = Function(V)
```
Comment (pbrubeck) [inline firedrake/ml/jax/fem_operator.py]:
```suggestion
        x_F = Function(V, val=np.asarray(x))
```

File: tests/firedrake/external_operators/test_external_operators.py [hunk 11]
```diff
@@ -131,8 +131,7 @@ def test_assemble(V, f):
 
     # -- dNdu(u, v, w; δN, uhat) (Adjoint model) -- #
     # Define a random cofunction on V* since the adjoint model maps from V* to V*
-    delta_N = Cofunction(V.dual())
-    delta_N.vector()[:] = rg.beta(V, 15, 30).dat.data_ro[:]
+    delta_N = Cofunction(V.dual()).assign(rg.beta(V, 15, 30))
```
Comment (pbrubeck) [inline tests/firedrake/external_operators/test_external_operators.py]:
```suggestion
    delta_N = rg.beta(V.dual(), 15, 30)
```

File: firedrake/ml/jax/fem_operator.py [hunk 12]
```diff
@@ -225,7 +222,6 @@ def from_jax(x: "jax.Array", V: Optional[WithGeometry] = None) -> Union[Function
...
-        x_F.vector().set_local(x)
+        x_F.dat.data_wo[...] = np.asarray(x)
```
Comment (pbrubeck) [inline firedrake/ml/jax/fem_operator.py]:
```suggestion
```

File: firedrake/ml/pytorch/fem_operator.py [hunk 13]
```diff
@@ -220,7 +217,6 @@ def from_torch(x, V=None):
...
-        x_F.vector().set_local(x)
+        x_F.dat.data_wo[...] = x.detach().numpy()
```
Comment (pbrubeck) [inline firedrake/ml/pytorch/fem_operator.py]:
```suggestion
```

File: firedrake/ml/pytorch/fem_operator.py (hunk 9 above)
Comment (pbrubeck) [reply firedrake/ml/pytorch/fem_operator.py]: Do we want a copy here? Or can we point to the same data?

File: tests/firedrake/external_operators/test_external_operators.py (hunk 11 above)
Comment (pbrubeck) [reply tests/firedrake/external_operators/test_external_operators.py]: I was not aware that we still could assign a `Function` to a `Cofunction`

File: firedrake/ml/pytorch/fem_operator.py (hunk 9 above)
Comment (connorjward) [reply firedrake/ml/pytorch/fem_operator.py]: No idea. This seems better though.

File: firedrake/adjoint_utils/function.py [hunk 14]
```diff
@@ -276,22 +276,15 @@ def _ad_dot(self, other, options=None):
 
     @staticmethod
     def _ad_assign_numpy(dst, src, offset):
-        range_begin, range_end = dst.vector().local_range()
+        range_begin, range_end = dst.dat.dataset.layout_vec.getOwnershipRange()
         m_a_local = src[offset + range_begin:offset + range_end]
-        dst.vector().set_local(m_a_local)
-        dst.vector().apply('insert')
-        offset += dst.vector().size()
+        dst.dat.data_wo[...] = m_a_local.reshape(dst.dat.data_wo.shape)
```
Comment (JHopeCollins) [inline firedrake/adjoint_utils/function.py:281]: In light of James' PR, this would unnecessarily increment the dat version.
```suggestion
        dst.dat.data_wo[...] = m_a_local.reshape(dst.dat.data_ro.shape)
```

File: tests/firedrake/adjoint/test_assemble.py [hunk 15]
```diff
@@ -1,12 +1,16 @@
...
-from numpy.random import rand
 from numpy.testing import assert_allclose
 
 from firedrake import *
 from firedrake.adjoint import *
 
 
+@pytest.fixture
+def rg():
+    return RandomGenerator(PCG64(seed=1234))
```
Comment (JHopeCollins) [inline tests/firedrake/adjoint/test_assemble.py:11]: This returns a deterministic sequence of pseudo-random numbers. However, if multiple tests call this fixture then each test will not get a deterministic subset of those numbers because the order in which the tests are run may change, for example if run with different number of splits or if more tests are added.

Usually you can just set `np.random.seed` at the beginning of each test then call `np.random.random_sample`.

File: tests/firedrake/adjoint/test_assignment.py [hunk 16]
```diff
@@ -4,10 +4,14 @@
 from firedrake.adjoint import *
 from checkpoint_schedules import SingleMemoryStorageSchedule
 
-from numpy.random import rand, default_rng
 from numpy.testing import assert_approx_equal, assert_allclose
 
 
+@pytest.fixture
+def rg():
+    return RandomGenerator(PCG64(seed=1234))
```
Comment (JHopeCollins) [inline tests/firedrake/adjoint/test_assignment.py:12]: Same as before, setting seed in each test will be deterministic

File: tests/firedrake/adjoint/test_hessian.py [hunk 17]
```diff
@@ -7,6 +7,11 @@
 rng = default_rng()
 
 
+@pytest.fixture
+def rg():
+    return RandomGenerator(PCG64(seed=1234))
```
Comment (JHopeCollins) [inline tests/firedrake/adjoint/test_hessian.py:12]: And here

**Review by JHopeCollins — DISMISSED** [review]

File: firedrake/adjoint_utils/function.py (hunk 14 above)
Comment (connorjward) [reply firedrake/adjoint_utils/function.py:281]: But we are immediately bumping the dat version anyway since we still need the `wo` option.

File: tests/firedrake/adjoint/test_assemble.py (hunk 15 above)
Comment (connorjward) [reply tests/firedrake/adjoint/test_assemble.py:11]: Actually this is OK I think. The fixture is regenerated for every test: https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#fixture-scopes.

File: firedrake/adjoint_utils/function.py (hunk 14 above)
Comment (JHopeCollins) [reply firedrake/adjoint_utils/function.py:281]: This would bump it twice wouldn't it?

File: tests/firedrake/adjoint/test_assemble.py (hunk 15 above)
Comment (JHopeCollins) [reply tests/firedrake/adjoint/test_assemble.py:11]: TIL!

File: firedrake/adjoint_utils/function.py (hunk 14 above)
Comment (connorjward) [reply firedrake/adjoint_utils/function.py:281]: But that doesn't matter. The dat version only matters to indicate that something has changed. We never care about the actual value.

**Review by JHopeCollins — APPROVED** [review]
LGTM, thanks
